parse5-utils.js

```diff
@@ -36,8 +35,160 @@ function setAttribute(element, name, value) {
   }
 }
 
+function hasTagName(name) {
+  return function(node) {
+    return node.tagName === name;
+  };
+}
+
+function hasClass(name) {
+  return function(node) {
+    var attr = getAttribute(node, 'class');
+    if (!attr) {
+      return false;
+    }
+    return attr.split(' ').indexOf(name) > -1;
+  };
+}
+
+// OR an array of predicates
+function OR(rules) {
+  return function(node) {
+    for (var i = 0; i < rules.length; i++) {
+      if (rules[i](node)) {
+        return true;
+      }
+    }
+    return false;
+  };
+}
+
+// AND an array of predicates
+function AND(rules) {
+  return function(node) {
+    for (var i = 0; i < rules.length; i++) {
+      if (!rules[i](node)) {
+        return false;
+      }
+    }
+    return true;
+  };
+}
+
+// negate an individual predicate, or a group with AND or OR
+function NOT(predicateFn) {
+  return function(node) {
+    return !predicateFn(node);
+  };
+}
+
+function hasAttr(attr) {
+  return function(node) {
+    return getAttributeIndex(node, attr) > -1;
+  };
+}
+
+function hasAttrValue(attr, value) {
+  return function(node) {
+    return getAttribute(node, attr) === value;
+  };
+}
+
+function isElement(node) {
+  return node.nodeName === node.tagName;
+}
+
+function query(node, predicate) {
+  if (isElement(node) && predicate(node)) {
+    return node;
+  }
+  var match = null;
+  if (node.childNodes) {
+    for (var i = 0; i < node.childNodes.length; i++) {
+      match = query(node.childNodes[i], predicate);
+      if (match) {
+        break;
+      }
+    }
+  }
+  return match;
+}
+
+function queryAll(node, predicate) {
+  var matches = [];
+  if (isElement(node) && predicate(node)) {
+    matches.push(node);
+  }
+  if (node.childNodes) {
+    for (var i = 0; i < node.childNodes.length; i++) {
+      matches = matches.concat(queryAll(node.childNodes[i], predicate));
```

Comment:
You could just push `matches` down the stack as an optional argument and get less allocations and concatenation